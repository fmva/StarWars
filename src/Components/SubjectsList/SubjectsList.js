import { Component } from 'react';
import React from 'react';
import { Icon, Input, Button } from 'antd';
import './SubjectsList.less';
import Resources from '../../Resources/Resources';
import Settings from '../../Resources/Settings';
import PropTypes from 'prop-types';

class SubjectsList extends Component {
  static propTypes = {
    /** array of subjects */
    subjects: PropTypes.array,
    /** name of object's field */
    titleSubjectField: PropTypes.string,
    /** title page */
    title: PropTypes.string,
    /** show/hide pagination button */
    isPaginationButton: PropTypes.bool,
    /** search value */
    searchValue: PropTypes.string,
    /** event for search field */
    onChangeSearch: PropTypes.func,
    /** handler to set pagination div */
    setPaginationDiv: PropTypes.func,
    /** pagination handler */
    onPagination: PropTypes.func,
    /** click item handler */
    onClickItem: PropTypes.func,
    /** reset search value */
    onResetSearchValue: PropTypes.func
  };

  constructor(props) {
    super(props);
    this.subjectDiv = React.createRef();

    this.onPaginationScroll = this.onPaginationScroll.bind(this);
  }

  componentDidMount() {
    this.props.setPaginationDiv(this.subjectDiv);
  }

  /**
   * pagination with scroll
   */
  onPaginationScroll() {
    const offsetHeight = this.subjectDiv.current.offsetHeight;
    const scrollHeight = this.subjectDiv.current.scrollHeight;
    const scrollTop = this.subjectDiv.current.scrollTop;
    if (scrollTop + offsetHeight > scrollHeight - Settings.getSettings('scrollOffset')) {
      this.props.onPagination();
    }
  }

  /**
   * render list of title
   * @param {Array} arr - array of subjects
   * @param {String} titleSubjectField - title of field
   * @return {Object}
   */
  renderMenu(arr, titleSubjectField) {
    const listItems = arr.map((item, key) => (
      <li key={`item_list_${key})`}>
        <div>
          <a
            href={`#list_${key}`}
            onClick={() => {
              this.props.onClickItem(key);
            }}
          >
            {item[titleSubjectField]}
          </a>
        </div>
      </li>
    ));

    if (listItems.length === 0) {
      return <div className="list-subjects-none">{Resources.getResource('none_list')}</div>;
    } else {
      return <ul className="list-subjects">{listItems}</ul>;
    }
  }

  render() {
    const subjects =
      this.props.subjects && this.props.subjects.length > 0 ? this.props.subjects : [];
    const titleSubjectField = this.props.titleSubjectField ? this.props.titleSubjectField : null;
    const title = this.props.title ? this.props.title : null;
    const onChangeSearch = this.props.onChangeSearch ? this.props.onChangeSearch : null;
    const isPaginationButton = this.props.isPaginationButton ? this.props.isPaginationButton : null;
    const onPagination = this.props.onPagination ? this.props.onPagination : null;
    const searchValue = this.props.searchValue ? this.props.searchValue : null;
    const onResetSearchValue = this.props.onResetSearchValue ? this.props.onResetSearchValue : null;

    return (
      <div ref={this.subjectDiv} className="list-subject-menu" onScroll={this.onPaginationScroll}>
        <h2>{title}</h2>
        <div className="search-input-block">
          <Input
            className="search-input"
            size="large"
            placeholder={Resources.getResource('search_button')}
            maxLength={60}
            onChange={onChangeSearch}
            value={searchValue}
            suffix={
              searchValue ? (
                <Icon
                  className="search-input-icon"
                  type="close-circle"
                  onClick={onResetSearchValue}
                />
              ) : (
                <span className="search-input-icon" />
              )
            }
          />
        </div>
        {this.renderMenu(subjects, titleSubjectField)}

        <div
          className="pagination-block"
          style={{ display: isPaginationButton ? 'block' : 'none' }}
        >
          <Button
            block
            type="primary"
            icon="down"
            size="large"
            title={Resources.getResource('show_more')}
            onClick={onPagination}
          />
        </div>
      </div>
    );
  }
}

export default SubjectsList;
