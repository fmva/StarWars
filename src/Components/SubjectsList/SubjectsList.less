@import "../../Less/overall.less";

.list-subject-menu{
  width: 70%;
  padding:10px;
  text-align: center;
  border-right:1px solid #cfd3e2;
  overflow: auto;
  order:1;
  @media @phone {
    border-right:none;
    width: 100%;
    height: 100%;
    padding-top: 62px;
  }
}

.search-input-block{
  margin: 0 auto;
  width: 70%;
}

.search-input{
  width: 100%;
  text-align: left;
}

.search-input-icon {
  width:100%;
}

ul.list-subjects{
  list-style-type: none;
  padding: @padding-menu-list;
  width: 65%;
  margin: 0 auto;
  @media @phone {
    width: 85%;
    height:100%;
  }
}

.list-subjects li{
  padding:1px 0;

}

.list-subjects div{
  border:1px solid #ececec;
  border-radius: 3px;
}

.list-subjects a{
  position: relative;
  display: block;
  padding: 10px;
  color: #3b3d44;
  font-weight: bold;
  text-decoration: none;
  box-shadow:inset 5px 0 #7b84ae;
  border:1px solid #b8c2d4;
  border-radius: 3px;
  background-color: #fff;
}

.list-subjects a:active{
  background-color:#b8c2d4;
}
.list-subjects a:hover{
  box-shadow:0 0 10px 0 #7b84ae;
}

.pagination-block{
  padding:5px 0;
  width:30%;
  margin: 0 auto;
  @media @phone {
    width:50%;
  }
}

.list-subjects-none {
  font-size: 2em;
  font-weight: bold;
  padding: 3px;
}
