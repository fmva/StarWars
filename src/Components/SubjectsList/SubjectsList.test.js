import SubjectsList from './SubjectsList';
import { shallow, mount } from 'enzyme';
import React from 'react';

const subjects = [
  {
    name: 'Luke Skywalker',
    height: '172',
    mass: '77',
    hair_color: 'blond',
    skin_color: 'fair',
    eye_color: 'blue',
    birth_year: '19BBY',
    gender: 'male'
  },
  {
    name: 'C-3PO',
    height: '167',
    mass: '75',
    hair_color: 'n/a',
    skin_color: 'gold',
    eye_color: 'yellow',
    birth_year: '112BBY',
    gender: 'n/a'
  }
];
let wrapper;
const titleSubjectField = 'name';
const title = 'People';
const onChangeSearch = jest.fn();
const setPaginationDiv = jest.fn();
const onPagination = jest.fn();
const onClickItem = jest.fn();
const onResetSearchValue = jest.fn();

describe('SubjectsList', () => {
  beforeAll(() => {
    wrapper = mount(
      <SubjectsList
        subjects={subjects}
        titleSubjectField={titleSubjectField}
        title={title}
        isPaginationButton={true}
        searchValue={'test'}
        onChangeSearch={onChangeSearch}
        setPaginationDiv={setPaginationDiv}
        onPagination={onPagination}
        onClickItem={onClickItem}
        onResetSearchValue={onResetSearchValue}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('Компонент SubjectsList с входными данными', () => {
    expect(wrapper.find('.list-subjects a').length).toBe(2);

    expect(
      wrapper
        .find('.list-subjects a')
        .at(0)
        .text()
    ).toBe('Luke Skywalker');
    expect(
      wrapper
        .find('.list-subjects a')
        .at(1)
        .text()
    ).toBe('C-3PO');
  });

  it(
    'Допустим пользователь вводит данные в поисковое поле, ' +
      'тогда должно сработать событие onChangeSearch',
    () => {
      wrapper.find('Input.search-input input.ant-input').simulate('change');
      expect(onChangeSearch.mock.calls.length).toBe(1);
    }
  );

  it(
    'Допустим пользователь нажимает на пункт списка, ' +
      'тогда должно сработать событие onClickItem',
    () => {
      wrapper
        .find('.list-subjects a')
        .at(0)
        .simulate('click');
      expect(onClickItem.mock.calls.length).toBe(1);
      wrapper
        .find('.list-subjects a')
        .at(1)
        .simulate('click');
      expect(onClickItem.mock.calls.length).toBe(2);
    }
  );

  it(
    'Допустим пользователь нажимает иконку очистки поискового поля, ' +
      'тогда должно сработать событие onResetSearchValue',
    () => {
      wrapper.find('Icon.search-input-icon i.search-input-icon').simulate('click');
      expect(onResetSearchValue.mock.calls.length).toBe(1);
    }
  );

  it('При формировании компонента, ' + 'должно сработать событие setPaginationDiv', () => {
    expect(setPaginationDiv.mock.calls.length).toBe(1);
  });
});
