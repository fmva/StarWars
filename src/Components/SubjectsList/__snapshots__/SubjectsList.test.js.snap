// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<SubjectsList
  isPaginationButton={true}
  onChangeSearch={[MockFunction]}
  onClickItem={[MockFunction]}
  onPagination={[MockFunction]}
  onResetSearchValue={[MockFunction]}
  searchValue="test"
  setPaginationDiv={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "current": <div
              class="list-subject-menu"
            >
              <h2>
                People
              </h2>
              <div
                class="search-input-block"
              >
                <span
                  class="search-input ant-input-affix-wrapper ant-input-affix-wrapper-lg"
                >
                  <input
                    class="ant-input ant-input-lg"
                    maxlength="60"
                    placeholder="Search"
                    type="text"
                    value="test"
                  />
                  <span
                    class="ant-input-suffix"
                  >
                    <i
                      aria-label="icon: close-circle"
                      class="anticon anticon-close-circle search-input-icon"
                      tabindex="-1"
                    >
                      <svg
                        aria-hidden="true"
                        class=""
                        data-icon="close-circle"
                        fill="currentColor"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M685.4 354.8c0-4.4-3.6-8-8-8l-66 .3L512 465.6l-99.3-118.4-66.1-.3c-4.4 0-8 3.5-8 8 0 1.9.7 3.7 1.9 5.2l130.1 155L340.5 670a8.32 8.32 0 0 0-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3L512 564.4l99.3 118.4 66 .3c4.4 0 8-3.5 8-8 0-1.9-.7-3.7-1.9-5.2L553.5 515l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z"
                        />
                        <path
                          d="M512 65C264.6 65 64 265.6 64 513s200.6 448 448 448 448-200.6 448-448S759.4 65 512 65zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                        />
                      </svg>
                    </i>
                  </span>
                </span>
              </div>
              <ul
                class="list-subjects"
              >
                <li>
                  <div>
                    <a
                      href="#list_0"
                    >
                      Luke Skywalker
                    </a>
                  </div>
                </li>
                <li>
                  <div>
                    <a
                      href="#list_1"
                    >
                      C-3PO
                    </a>
                  </div>
                </li>
              </ul>
              <div
                class="pagination-block"
                style="display: block;"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-lg ant-btn-icon-only ant-btn-block"
                  title="Show More"
                  type="button"
                >
                  <i
                    aria-label="icon: down"
                    class="anticon anticon-down"
                  >
                    <svg
                      aria-hidden="true"
                      class=""
                      data-icon="down"
                      fill="currentColor"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </i>
                </button>
              </div>
            </div>,
          },
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  subjects={
    Array [
      Object {
        "birth_year": "19BBY",
        "eye_color": "blue",
        "gender": "male",
        "hair_color": "blond",
        "height": "172",
        "mass": "77",
        "name": "Luke Skywalker",
        "skin_color": "fair",
      },
      Object {
        "birth_year": "112BBY",
        "eye_color": "yellow",
        "gender": "n/a",
        "hair_color": "n/a",
        "height": "167",
        "mass": "75",
        "name": "C-3PO",
        "skin_color": "gold",
      },
    ]
  }
  title="People"
  titleSubjectField="name"
>
  <div
    className="list-subject-menu"
    onScroll={[Function]}
  >
    <h2>
      People
    </h2>
    <div
      className="search-input-block"
    >
      <Input
        className="search-input"
        disabled={false}
        maxLength={60}
        onChange={[MockFunction]}
        placeholder="Search"
        size="large"
        suffix={
          <Icon
            className="search-input-icon"
            onClick={[MockFunction]}
            type="close-circle"
          />
        }
        type="text"
        value="test"
      >
        <span
          className="search-input ant-input-affix-wrapper ant-input-affix-wrapper-lg"
        >
          <input
            className="ant-input ant-input-lg"
            disabled={false}
            maxLength={60}
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Search"
            size="large"
            style={null}
            type="text"
            value="test"
          />
          <span
            className="ant-input-suffix"
          >
            <Icon
              className="search-input-icon"
              onClick={[MockFunction]}
              type="close-circle"
            >
              <LocaleReceiver
                componentName="Icon"
              >
                <i
                  aria-label="icon: close-circle"
                  className="anticon anticon-close-circle search-input-icon"
                  onClick={[MockFunction]}
                  tabIndex={-1}
                >
                  <IconReact
                    className=""
                    type="close-circle-o"
                  >
                    <svg
                      aria-hidden="true"
                      className=""
                      data-icon="close-circle"
                      fill="currentColor"
                      height="1em"
                      key="svg-close-circle"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M685.4 354.8c0-4.4-3.6-8-8-8l-66 .3L512 465.6l-99.3-118.4-66.1-.3c-4.4 0-8 3.5-8 8 0 1.9.7 3.7 1.9 5.2l130.1 155L340.5 670a8.32 8.32 0 0 0-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3L512 564.4l99.3 118.4 66 .3c4.4 0 8-3.5 8-8 0-1.9-.7-3.7-1.9-5.2L553.5 515l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z"
                        key="svg-close-circle-svg-0"
                      />
                      <path
                        d="M512 65C264.6 65 64 265.6 64 513s200.6 448 448 448 448-200.6 448-448S759.4 65 512 65zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
                        key="svg-close-circle-svg-1"
                      />
                    </svg>
                  </IconReact>
                </i>
              </LocaleReceiver>
            </Icon>
          </span>
        </span>
      </Input>
    </div>
    <ul
      className="list-subjects"
    >
      <li
        key="item_list_0)"
      >
        <div>
          <a
            href="#list_0"
            onClick={[Function]}
          >
            Luke Skywalker
          </a>
        </div>
      </li>
      <li
        key="item_list_1)"
      >
        <div>
          <a
            href="#list_1"
            onClick={[Function]}
          >
            C-3PO
          </a>
        </div>
      </li>
    </ul>
    <div
      className="pagination-block"
      style={
        Object {
          "display": "block",
        }
      }
    >
      <Button
        block={true}
        ghost={false}
        icon="down"
        loading={false}
        onClick={[MockFunction]}
        size="large"
        title="Show More"
        type="primary"
      >
        <Wave>
          <button
            className="ant-btn ant-btn-primary ant-btn-lg ant-btn-icon-only ant-btn-block"
            onClick={[Function]}
            title="Show More"
            type="button"
          >
            <Icon
              type="down"
            >
              <LocaleReceiver
                componentName="Icon"
              >
                <i
                  aria-label="icon: down"
                  className="anticon anticon-down"
                >
                  <IconReact
                    className=""
                    type="down-o"
                  >
                    <svg
                      aria-hidden="true"
                      className=""
                      data-icon="down"
                      fill="currentColor"
                      height="1em"
                      key="svg-down"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        key="svg-down-svg-0"
                      />
                    </svg>
                  </IconReact>
                </i>
              </LocaleReceiver>
            </Icon>
          </button>
        </Wave>
      </Button>
    </div>
  </div>
</SubjectsList>
`;
